hermes_core/util/cdf.py

```diff
@@ -0,0 +1,612 @@
+    data structure that can be manipulated in memory. The internal `dict` used `self.global_attrs`
+    is formatted in the following way.
+
+    ```py
+    self.global_attrs = {
+        'global_attribute_name1': 'global_attribute_value1',
+        'global_attribute_name2': 'global_attribute_value2',
+    }
+
+    ```
+    """
+
+    def __init__(self):
+
+        # Current Working CDF File
+        self.target_cdf_file = None
+
+        # Data Validation, Complaiance, Derived Attributes
+        self.schema = {}
+        # Load Default schema Data
+        self._load_default_schema_data()
+
+        # Intermediate Data Structure for Global / Metadata Attributes
+        self.global_attrs = {}
+        # Load Default Global Attributes
+        self._load_default_attributes()
+
+        # Intermedaite Data Structure for Variable Data
+        self.variable_data = {}
+
+        # Intermediate Data Structure for Variable Attributes
+        self.variable_attrs = {}
+
+    def __del__(self):
+        """
+        Deconstructor for CDFWriter class.
+        """
+        if self.target_cdf_file:
+            self.target_cdf_file.close()
+
+    def __len__(self):
+        """
+        Function to get the number of variable data members in the CDFWriter class.
+        """
+        return len(self.variable_data.keys())
+
+    def __repr__(self):
+        """
+        Returns a representation of the CDFWriter class.
+        """
+        return self.__str__()
+
+    def __str__(self):
+        """
+        Returns a string representation of the CDFWriter class.
+        """
+        str_repr = (
+            f"CDFWriter: Converted: {self.target_cdf_file is not None}"
+            f"\nGlobal Attrs: {self.global_attrs}"
+            f"\nVariable Data: {self.variable_data}"
+            f"\nVariable Attributes: {self.variable_attrs}"
+        )
+        return str_repr
+
+    def __getitem__(self, name):
+        """
+        Function to get a data variable contained in the CDFWriter class.
+        """
+        if name not in self.variable_data:
+            raise KeyError(f"CDFWriter does not contain data variable {name}")
+        # Get the Data and Attrs for the named variable
+        var_data = self.variable_data[name]
+        var_attrs = self.variable_attrs[name] if name in self.variable_attrs else {}
+        return (var_data, var_attrs)
+
+    def __setitem__(self, name, data):
+        """
+        Function to set a data variable conained in the CDFWriter class.
+        """
+        # Set the Data for the named variable
+        self.variable_data[name] = data
+        # Set the Attrs for the named variable
+        if name not in self.variable_attrs:
+            self.variable_attrs[name] = {}
+
+    def __contains__(self, name):
+        """
+        Function to see whether a data variable is in the CDFWriter class.
+        """
+        return name in self.variable_data
+
+    def __delitem__(self, name):
```

Comment:
Given that you are defining all of these you may consider subclassing `list`, or using a `dataclass`.

Reply:
The goal of overriding these was to make an easier interface for getting at the data, treating the class object as dict like. we could do:
```py
a['variable_name']
```
rather than 
```py
a.data['variable_name']
````